src/rules/restrict-dependency-ranges.ts

```diff
@@ -0,0 +1,293 @@
+		}
+	}
+
+	return version.replace(
+		/^(workspace:)?(\^|~|<=?|>=?)?/,
+		`$1${SYMBOLS[rangeType]}`,
+	);
+};
+
+/**
+ * Check if the version is in a form that this rule supports.
+ */
+const isVersionSupported = (version: string): boolean => {
+	if (version === "*" || /^workspace:[*^~]$/.test(version)) {
+		return true;
+	}
+
+	const rawVersion = version.replace(/^workspace:/, "");
+	return !!semver.valid(rawVersion) || !!semver.validRange(rawVersion);
+};
+
+const capitalize = (str: string): string => {
+	return str.charAt(0).toUpperCase() + str.slice(1);
+};
+
+export const rule = createRule<Options>({
+	create(context) {
+		// Bail early if no options were provided
+		if (!context.options[0]) {
+			return {};
+		}
+
+		const optionsArray = Array.isArray(context.options[0])
+			? [...context.options[0]]
+			: [context.options[0]];
+
+		// Reverse the array, so that subsequent options override previous ones
+		optionsArray.reverse();
+
+		return {
+			"Program > JSONExpressionStatement > JSONObjectExpression > JSONProperty[key.type=JSONLiteral][value.type=JSONObjectExpression]"(
+				node: JsonAST.JSONProperty & {
+					key: JsonAST.JSONStringLiteral;
+					value: JsonAST.JSONObjectExpression;
+				},
+			) {
+				const dependencyType = node.key.value;
+
+				// If this isn't a group of dependencies, skip it entirely
+				if (!DEPENDENCY_TYPES.includes(dependencyType)) {
+					return;
+				}
+
+				// Loop through all dependencies in the group
+				for (const property of node.value.properties) {
+					// If either the key or value aren't strings, this isn't a valid dependency, so move on.
+					if (
+						!isJSONStringLiteral(property.key) ||
+						!isJSONStringLiteral(property.value)
+					) {
+						continue;
+					}
+
+					const name = property.key.value;
+					const version = property.value.value;
+
+					// Check to see if the version is in a format we support
+					if (!isVersionSupported(version)) {
+						continue;
+					}
+
+					const isPinned =
+						!!semver.parse(version) || version === "workspace:*";
+					const isTildeRange =
+						(!!semver.validRange(version) &&
+							version.startsWith("~")) ||
+						version.startsWith("workspace:~");
+					const isCaretRange =
+						(!!semver.validRange(version) &&
+							version.startsWith("^")) ||
+						version.startsWith("workspace:^");
+
+					// Loop through all options, and evaluate each of them for this dependency
+					for (const options of optionsArray) {
+						// Skip these options if they have any conditions that match
+						// the current dependency.
+						if (
+							options.forDependencyTypes &&
+							!options.forDependencyTypes.includes(dependencyType)
+						) {
+							continue;
+						}
+						if (options.forPackages) {
+							let isMatch = false;
+							for (const packageNamePattern of options.forPackages) {
+								if (new RegExp(packageNamePattern).test(name)) {
+									isMatch = true;
+									break;
+								}
+							}
```

Comment:
[Style] Nit: we can avoid the extra mutable variable with a little `.some`:

```suggestion
							const isMatch = options.forPackages.some(
								(packageNamePattern) =>
									new RegExp(packageNamePattern).test(name),
							);
```

Reply:
Updated